### ISSUE TITLE
Mac problem: Not using the bundled JRE?

### ISSUE THREAD

**Issue by charxie** [libgdx/packr#179]
We tried to bundled Java 8 but it seems it doesn't pick up the JRE.  The command returned no error and the JRE was successfully copied into the Resources folder. Did anyone have the same issue?

**Comment by karlsabo** [comment]
Which command? Are you trying to run your packed app? If so have you tried running from the command line with verbose output turned on? What other steps have you tried to debug the issue?

**Comment by lichenglu** [comment]
Thanks for the fast reply @karlsabo. I am working with @charxie.

We are both testing in the MacOS environment. For your information. @charxie is using Oracle JDK 8 and the packed app works fine on his computer. However, it is not working for me and I am using OpenJDK 8. We have tried the following ways to run the executable

1. Double click `Contents -> MacOS -> executable file`
2. Make the output folder an App by adding the `.app` extension

It works if I run the jar file direction with `./jre/bin/java [myapp].jar` under `Resources`. That's why we are suspecting the executable in `MacOS` is not using the shipped JRE.

Any thoughts? Thank you!

**Comment by karlsabo** [comment]
Open a terminal and run
`outdir/Contents/MacOS/executable-name --cli --verbose --`

Where outdir is the output directory you specified when running packr-all.jar and executable-name is the name you specified for that too.

That should give some good hints as to what's going on.

**Comment by lichenglu** [comment]
Here are the outputs:

```
Changing working directory to outdir/Contents/Resources ...
Loading JVM runtime library ...
fileSearch found libjli! filename=jre/lib/jli/libjli.dylib, lastFileSearch=
Loading libjli=outdir/Contents/Resources/jre/lib/jli/libjli.dylib
Passing VM options ...
isZgcSupported()=1, hasJsonValue(jsonRoot, "useZgcIfSupportedOs", sajson::TYPE_TRUE)=0
  # -Xmx512M
Passing VM options:
  -Xmx512M
Creating Java VM ...
Passing command line arguments ...
Loading JAR file ...
Adding 1 classpaths ...
  # aladdin.jar
Invoking static org.aladdin.App.main() function ...
Initiating...
Checking single instance...
1.8.0_162, x86_64
Path = ./lib/jogl/native/mac-universal
Log folder: log
Initialization phase 2 done.
Time = 1.478569135
Checked for an exception from the main method, exceptionOccurred=0
```

Here is the error:

```
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'NSWindow drag regions should only be invalidated on the Main Thread!'
```

It does seem that the JRE is being used?

**Comment by karlsabo** [comment]
Yes, it's using the bundled JRE, it would take some effort to get it to use anything else. 

I bet your missing `-XstartOnFirstThread` as a JVM option.

**Comment by lichenglu** [comment]
Thank you! One last question, do we do like this?

```
java -jar packr-all.jar \
     --vmargs Xmx1G -XstartOnFirstThread \
     --...
```

Or we use use XstartOnFirstThread without the `-` (dash)? Because we have tried it without the dash, and the program will just hang there forever for both of us.

**Comment by karlsabo** [comment]
Remove the leading dash. 

Run with verbose and see if it gets any further. Also have you tried attaching a debugger or profiler such as visualVM to see what threads are hanging and where?

**Comment by charxie** [comment]
> Remove the leading dash.
> 
> Run with verbose and see if it gets any further. Also have you tried attaching a debugger or profiler such as visualVM to see what threads are hanging and where?

When we did this,

**Comment by charxie** [comment]
> Remove the leading dash.
> 
> Run with verbose and see if it gets any further. Also have you tried attaching a debugger or profiler such as visualVM to see what threads are hanging and where?

 Did you mean run packr with verbose? We tried that but there was no more message.

**Comment by karlsabo** [comment]
Did you get it working?

**Comment by charxie** [comment]
Nope. I accidentally closed this issue. There seems to be no way to undo that. So I reopened it as a new issue. Sorry about that.

**Comment by charxie** [comment]
Seems there is a way to reopen it. So I moved my new posts here:

Sorry, I accidentally hit the wrong button and closed the issue by mistake.

This is the command we used:

java -jar packr-all-3.0.1.jar --platform mac --jdk /Library/Java/JavaVirtualMachines/jdk1.8.0_162.jdk --executable Aladdin --classpath aladdin.jar --mainclass org.aladdin.App --vmargs Xmx512M XstartOnFirstThread --resources lib --output ala --verbose

And the output is:

Copying executable ...
Copying classpath(s) ...
Unpacking JRE ...
Copying resources ...
Done!

Thank you so much in advance for your help!

**Comment by karlsabo** [comment]
I meant to run the packr launcher exe in verbose mode.

`ala/Contents/MacOS/Aladdin --cli --verbose --`

If you have logging, you might be able to figure out where it's hanging, otherwise use [jstack](https://docs.oracle.com/javase/8/docs/technotes/tools/unix/jstack.html) or [VisualVM](https://visualvm.github.io/). jstack will give you a thread dump in the console, VisualVM is a bit nicer and easier to use.

**Comment by charxie** [comment]
This is what it outputs:

```
ala/Contents/MacOS/Aladdin --cli --verbose --
Changing working directory to /Users/ccuser/IdeaProjects/aladdin/out/artifacts/aladdin/ala/Contents/Resources ...
Loading JVM runtime library ...
fileSearch found libjli! filename=jre/lib/jli/libjli.dylib, lastFileSearch=
Loading libjli=/Users/ccuser/IdeaProjects/aladdin/out/artifacts/aladdin/ala/Contents/Resources/jre/lib/jli/libjli.dylib
Passing VM options ...
isZgcSupported()=1, hasJsonValue(jsonRoot, "useZgcIfSupportedOs", sajson::TYPE_TRUE)=0
  # -Xmx512M
  # -XstartOnFirstThread
Passing VM options:
  -Xmx512M
  -XstartOnFirstThread
Starting JVM on main thread (-XstartOnFirstThread found) ...
Creating Java VM ...
Passing command line arguments ...
Loading JAR file ...
Adding 1 classpaths ...
  # aladdin.jar
Invoking static org.aladdin.App.main() function ...
Initiating...
Checking single instance...
1.8.0_162, x86_64
Path = ./lib/jogl/native/mac-universal
```

**Comment by karlsabo** [comment]
That looks good, definitely going to need to run jstack or VisualVM, the VM is up and running the code in aladdin.jar. Maybe it's hung on some JOGL call.

**Comment by charxie** [comment]
This app indeed uses Jogl. We will check. Is the VM argument a must in order to get it to work? How do you explain that it works on the computer that originally built it?

**Comment by karlsabo** [comment]
Graphics APIs usually have to be called from the main threads. It's required on macOS if you're doing any UI stuff. 

You can search for that VM parameter, it's hard to find any good info on it. https://www.eclipse.org/swt/faq.php#swtawtosxmore

My guess is your app is hung on a graphics or windowing call.

**Comment by charxie** [comment]
Thanks for your help! We will try again tomorrow!

**Comment by charxie** [comment]
You are right about the windowing call. The problem is that Mac has made a rule that is incompatible with Java Swing, which requires the initialization of Swing components to be in the AWT thread, not the main thread. Not sure how this problem can be solved.

**Comment by karlsabo** [comment]
You could try putting everything in `SwingUtilities.invokeLater` calls. It's really hard to mix LWJGL 2 or JOGL with Swing.

Swing doesn't play nicely with startOnFirstThread but LWJGL, GLFW and afaik JOGL all require it.

**Comment by charxie** [comment]
My problem now boils down to the fact that the runnable in EventQueue.invokeLater (equivalent to SwingUtilities.invokeLater) doesn't run from the main thread. It just hangs there. Any idea?

**Comment by karlsabo** [comment]
I don't have any other ideas, but someone in the libGDX community might <https://libgdx.badlogicgames.com/community.html>.
